torchchat/model_dist.py

```diff
@@ -107,22 +88,38 @@ def setup_caches(self, max_batch_size, max_seq_length):
         )
         self.register_buffer("causal_mask", causal_mask, persistent=True)
 
+    def distribute(self, device_mesh: DeviceMesh):
+        if hasattr(self, "tok_embeddings"):
+            parallelize_module(
+                self.tok_embeddings,
+                device_mesh,
+                RowwiseParallel(input_layouts=Replicate()),
+            )
+        for layer in self.layers.values():
+            layer.distribute(device_mesh)
+        # TODO (kwen2501) : parallelize these
+        """
+        parallelize_module(self.norm, device_mesh, ...)
+        parallelize_module(self.output, device_mesh, ...)
+        """
+
     def forward(self, x: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
         assert self.freqs_cis is not None, "Caches must be initialized first"
         if input_pos is None:
             input_pos = torch.arange(x.shape[1], device=x.device, dtype=torch.long)
         mask = self.causal_mask[None, None, input_pos]
         freqs_cis = self.freqs_cis[input_pos]
 
-        if self.stage_idx == 0:
-            x: DTensor = self.tok_embeddings(x)
-            # TODO: sequence parallelize this
+        if hasattr(self, "tok_embeddings"):
```

Comment:
minor - is it possible to just check stage_idx as before since embeddings will always be on stage0, or alternatively do this hasattr in a post init type of deal and set a self.hasnorm bool property so we're not having to run through all the class dicts * every hasattr check, over and over on every token coming through.
My (minor) concern is that having all these attr checks in a forward could add up to a bit of perf hit in the larger models vs a bool or int check like before.  
More important to ensure things work first so this is fine but maybe just flag a TODO for perf streamlining.

Reply:
Makes sense. I am changing it to `if self.tok_embeddings:`. Correspondingly, at init, `None` will be assigned if not stage 0.